File: java/ninety-nine/src/main/java/com/rbleggi/ninetynine/P01.java
package com.rbleggi.ninetynine;

import java.util.List;
import java.util.NoSuchElementException;

public class P01 {
    public static <A> A last(List<A> items) {
        System.out.println("Recursion step: " + items);
        if (items.isEmpty()) {
            throw new NoSuchElementException("last of empty list");
        }
        if (items.size() == 1) {
            return items.get(0);
        }
        return last(items.subList(1, items.size()));
    }

    public static void main(String[] args) {
        System.out.println("Find the last element of a list.");
        var numbers = List.of(1, 1, 2, 3, 5, 8);
        System.out.println("The list is: " + numbers);
        var lastElement = last(numbers);
        System.out.println("The last element is: " + lastElement);
    }
}

File: kotlin/ninety-nine/src/main/kotlin/com/rbleggi/ninetynine/P01.kt
package com.rbleggi.ninetynine

object P01 {
    tailrec fun <A> last(items: List<A>): A {
        println("Recursion step: $items")
        return when {
            items.isEmpty() -> throw NoSuchElementException("last of empty list")
            items.size == 1 -> items[0]
            else -> last(items.drop(1))
        }
    }
}

fun mainP01() {
    println("Find the last element of a list.")
    val numbers = listOf(1, 1, 2, 3, 5, 8)
    println("The list is: $numbers")
    val lastElement = P01.last(numbers)
    println("The last element is: $lastElement")
}
